core/ast_node.js

```diff
@@ -0,0 +1,656 @@
+      var newWsCoordinate = new goog.math.Coordinate(newX, this.wsCoordinate_.y);
+      newAstNode = Blockly.ASTNode.createWorkspaceNode(this.location_,
+          newWsCoordinate);
+      break;
+
+    case Blockly.ASTNode.types.STACK:
+      var nextTopBlock = this.navigateBetweenStacks_(true);
+      newAstNode = Blockly.ASTNode.createStackNode(nextTopBlock);
+      break;
+
+    case Blockly.ASTNode.types.OUTPUT:
+      newAstNode = Blockly.ASTNode.createBlockNode(this.location_.sourceBlock_);
+      break;
+
+    case Blockly.ASTNode.types.FIELD:
+      var parentInput = this.location_.getParentInput();
+      newAstNode = this.findNextForField_(this.location_, parentInput);
+      break;
+
+    case Blockly.ASTNode.types.INPUT:
+      var inputConnection = this.location_.connection;
+      newAstNode = this.findNextForInput_(inputConnection, this.location_);
+      break;
+
+    case Blockly.ASTNode.types.BLOCK:
+      var nextConnection = this.location_.nextConnection;
+      newAstNode = Blockly.ASTNode.createConnectionNode(nextConnection);
+      break;
+
+    case Blockly.ASTNode.types.PREVIOUS:
+      var output = this.location_.outputConnection;
+      if (output) {
+        newAstNode = Blockly.ASTNode.createConnectionNode(output);
+      } else {
+        var srcBlock = this.location_.sourceBlock_;
+        newAstNode = Blockly.ASTNode.createBlockNode(srcBlock);
+      }
+      break;
+
+    case Blockly.ASTNode.types.NEXT:
+      if (this.location_.targetBlock()) {
+        newAstNode = this.findTopASTConnection_(this.location_.targetBlock());
+      }
+      break;
+  }
+
+  return newAstNode;
+};
+
+/**
+ * Find the element one level below and all the way to the left of the current
+ * location.
+ * @return {?Blockly.ASTNode} An ast node that wraps the next field, connection,
+ * workspace, or block.
+ */
+Blockly.ASTNode.prototype.in = function() {
+  if (!this.location_) {return null;}
+  var newAstNode = null;
+
+  switch (this.type_) {
+    case Blockly.ASTNode.types.WORKSPACE:
+      var firstBlock = this.location_.getTopBlocks()[0];
+      newAstNode = this.findTopASTConnection_(firstBlock);
+      newAstNode.type_ = Blockly.ASTNode.types.STACK;
+      break;
+
+    case Blockly.ASTNode.types.STACK:
+      newAstNode = this.findTopASTConnection_(this.location_);
+      break;
+
+    case Blockly.ASTNode.types.BLOCK:
+      var inputs = this.location_.inputList;
+      if (inputs && inputs.length > 0) {
+        var newParentInput = inputs[0];
+        var fieldNode =
+            this.findNextEditableField_(this.location_, newParentInput, true);
+        if (fieldNode) {
+          newAstNode = fieldNode;
+        } else {
+          newAstNode = Blockly.ASTNode.createInputNode(newParentInput);
+        }
+      }
+      break;
+
+    case Blockly.ASTNode.types.INPUT:
+      var inputConnection = this.location_.connection;
+      if (inputConnection.targetBlock()) {
```

Comment:
you can use inputConnection.targetConnection since that will be the top connection on the block.